pkg/operator/ceph/object/objectstore.go

```diff
@@ -56,129 +62,448 @@ type idType struct {
+
+	zoneIsMaster, err := checkZoneIsMaster(objContext)
+	if err != nil {
+		return err
+	}
+	zoneGroupIsMaster := false
 
-		// create the zonegroup if it doesn't exist yet
-		output, err = runAdminCommand(context, "zonegroup", "get")
+	if zoneIsMaster {
+		// endpoints that are part of a master zone are supposed to be the endpoints for a zone group
+		_, err := RunAdminCommandNoMultisite(objContext, "zonegroup", "modify", realmArg, zoneGroupArg, endpointArg)
 		if err != nil {
-			updatePeriod = true
-			output, err = runAdminCommand(context, "zonegroup", "create", "--master", endpointArg)
-			if err != nil {
-				return errors.Wrapf(err, "failed to create rgw zonegroup for %q", context.Name)
-			}
+			return errors.Wrapf(err, "failed to add object store %q in rgw zone group %q", objContext.Name, objContext.ZoneGroup)
 		}
+		logger.Debugf("endpoints for zonegroup %q are now %q", objContext.ZoneGroup, zoneEndpoints)
 
-		zoneGroupID, err := decodeID(output)
+		// check if zone group is master only if zone is master for creating the system user
+		zoneGroupIsMaster, err = checkZoneGroupIsMaster(objContext)
 		if err != nil {
-			return errors.Wrap(err, "failed to parse zone group id")
+			return errors.Wrapf(err, "failed to find out whether zone group %q in is the master zone group", objContext.ZoneGroup)
 		}
+	}
+	_, err = RunAdminCommandNoMultisite(objContext, "zone", "modify", realmArg, zoneGroupArg, zoneArg, endpointArg)
+	if err != nil {
+		return errors.Wrapf(err, "failed to add object store %q in rgw zone %q", objContext.Name, objContext.Zone)
+	}
+	logger.Debugf("endpoints for zone %q are now %q", objContext.Zone, zoneEndpoints)
+
+	// the period will help notify other zones of changes if there are multi-zones
+	_, err = RunAdminCommandNoMultisite(objContext, "period", "update", "--commit", realmArg, zoneGroupArg, zoneArg)
+	if err != nil {
+		return errors.Wrap(err, "failed to update period")
+	}
+	logger.Infof("added object store %q to realm %q, zonegroup %q, zone %q", objContext.Name, objContext.Realm, objContext.ZoneGroup, objContext.Zone)
 
-		// create the zone if it doesn't exist yet
-		output, err = runAdminCommand(context, "zone", "get", zoneArg)
+	// create system user for realm for master zone in master zonegorup for multisite scenario
+	if zoneIsMaster && zoneGroupIsMaster {
+		err = createSystemUser(objContext, namespace)
 		if err != nil {
-			updatePeriod = true
-			output, err = runAdminCommand(context, "zone", "create", "--master", endpointArg, zoneArg)
+			return err
+		}
+	}
+
+	return nil
+}
+
+func createSystemUser(objContext *Context, namespace string) error {
+	uid := objContext.Realm + "-system-user"
+	uidArg := fmt.Sprintf("--uid=%s", uid)
+	realmArg := fmt.Sprintf("--rgw-realm=%s", objContext.Realm)
+	zoneGroupArg := fmt.Sprintf("--rgw-zonegroup=%s", objContext.ZoneGroup)
+	zoneArg := fmt.Sprintf("--rgw-zone=%s", objContext.Zone)
+
+	output, err := RunAdminCommandNoMultisite(objContext, "user", "info", uidArg)
```

Comment:
As a general coding structure, it's preferred to return sooner if possible to avoid deep code blocks. For example:
```
output, err := RunAdminCommandNoMultisite(objContext, "user", "info", uidArg)
if err == nil {
  logger.Debugf("realm system user %q has already been created", uid)
  return nil
}
if code, ok := exec.ExitStatus(err); ok && code != int(syscall.EINVAL) {
  return errors.Wrapf(err, "radosgw-admin user info for system user failed with code %d and output %q", code, output)
}
// Insert code to create the system user
```